packages/react-form-hooks/src/hooks/list/list.ts

```diff
@@ -0,0 +1,264 @@
+ *         </label>
+ *         {field.title.error && <p>{field.title.error}</p>}
+ *         <label htmlFor={`description-${index}`}>
+ *           description{' '}
+ *           <input
+ *             id={`description-${index}`}
+ *             name={`description-${index}`}
+ *             value={fields.description.value}
+ *             onChange={fields.description.onChange}
+ *             onBlur={fields.description.onBlur}
+ *           />
+ *         </label>
+ *         {field.description.error && <p>{field.description.error}</p>}
+ *       </li>
+ *      ))}
+ *    </ul>
+ *  );
+ *}
+ *```
+ *
+ * If using `@shopify/polaris` or other custom components that support `onChange`, `onBlur`, `value`, and `error` props then
+ * you can accomplish the above more tersely by using the ES6 `spread` (...) operator.
+ *
+ * ```tsx
+ * function MyComponent() {
+ *  const title = useField([{title: '', description: ''}, {title: '', description: ''}]);
+ *
+ *  return (
+ *    <ul>
+ *     {variants.map((fields, index) => (
+ *       <li key={index}>
+ *         <TextField label="title" name={`title${index}`} {...fields.title} />
+ *         <TextField
+ *            label="description"
+ *            id={`description${index}`}
+ *            {...fields.description}
+ *         />
+ *       </li>
+ *      ))}
+ *    </ul>
+ *   );
+ * }
+ * ```
+ *
+ * @param config - A configuration object specifying both the value and validation config.
+ * @param validationDependencies - An array of dependencies to use to decide when to regenerate validators.
+ * @returns A list of dictionaries of `Field` objects representing the state of your input. It also includes functions to manipulate that state. Generally, you will want to pass these callbacks down to the component or components representing your input.
+ *
+ * @remarks
+ * **Reinitialization:** If the `list` property of the field configuration changes between calls to `useList`,
+ * the field will be reset to use it as it's new default value.
+ *
+ * **Imperative methods:** The returned `Field` objects contains a number of methods used to imperatively alter their state.
+ * These should only be used as escape hatches where the existing hooks and components do not make your life easy,
+ * or to build new abstractions in the same vein as `useForm`, `useSubmit` and friends.
+ */
+export function useList<Item extends object>(
+  {list, validates = {}}: FieldListConfig<Item>,
+  validationDependencies: unknown[] = [],
+): FieldDictionary<Item>[] {
+  const [state, dispatch] = useListReducer(list);
+  useEffect(
+    () => {
+      dispatch(reinitializeAction(list));
+    },
+    [list, dispatch],
+  );
+
+  const validationConfigs = useMemo(
+    () =>
+      mapObject<NormalizedValidationDictionary<any>>(
+        validates,
+        normalizeValidation,
+      ),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [validates, ...validationDependencies],
+  );
+
+  const handlers = useMemo(
```

Comment:
Originally this thing was based on `useField` in a loop but we can't actually do that because of how hooks work. As a result we have to do some weird stuff with building objects in a loop that call reducer actions.